include/oneapi/dpl/pstl/utils.h

```diff
@@ -818,30 +822,109 @@ __shars_upper_bound(_Acc __acc, _Size __first, _Size __last, const _Value& __val
+{
+#if _ONEDPL_CPP20_CONCEPTS_PRESENT
+template <typename _Iterator1, typename _Iterator2>
+struct __is_equality_comparable : std::bool_constant<std::equality_comparable_with<_Iterator1, _Iterator2>>
+{
+};
+#else
+template <typename _Iterator, typename = void>
+struct __base_iterator_type : std::true_type
+{
+    using __type = std::decay_t<_Iterator>;
+};
+
+template <typename _Iterator>
+struct __base_iterator_type<_Iterator, std::void_t<decltype(std::declval<std::decay_t<_Iterator>>().base())>>
+    : std::false_type
+{
+    using __type = decltype(std::declval<std::decay_t<_Iterator>>().base());
+};
+
+template <typename _Iterator, typename = void>
+struct __iterator_value_type
+{
+    using __type = void;
+};
+
+template <typename _Iterator>
+struct __iterator_value_type<_Iterator, std::void_t<typename std::iterator_traits<std::decay_t<_Iterator>>::value_type>>
+{
+    using __type = typename std::iterator_traits<std::decay_t<_Iterator>>::value_type;
+};
+
 template <typename _Iterator1, typename _Iterator2, typename = void>
-struct __is_equality_comparable : std::false_type
+struct __has_same_value_types : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __has_same_value_types<_Iterator1, _Iterator2,
+                              std::enable_if_t<std::is_same_v<typename __iterator_value_type<_Iterator1>::__type,
+                                                              typename __iterator_value_type<_Iterator2>::__type>>>
+    : std::true_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct __has_equality_op : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __has_equality_op<_Iterator1, _Iterator2,
+                         std::void_t<decltype(std::declval<_Iterator1>() == std::declval<_Iterator2>())>>
+    : std::true_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct __is_equality_self_comparable : std::false_type
 {
 };
 
-// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-struct __is_equality_comparable<
-    _Iterator1, _Iterator2,
-    std::void_t<decltype(::std::declval<::std::decay_t<_Iterator1>>() == ::std::declval<::std::decay_t<_Iterator2>>())>>
+struct __is_equality_self_comparable<_Iterator1, _Iterator2,
+                                     std::enable_if_t<std::conjunction_v<__has_same_value_types<_Iterator1, _Iterator2>,
+                                                                         __has_equality_op<_Iterator1, _Iterator2>>>>
     : std::true_type
 {
 };
 
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct __is_equality_comparable : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __is_equality_comparable<_Iterator1, _Iterator2,
+                                std::enable_if_t<__is_equality_self_comparable<_Iterator1, _Iterator2>::value>>
```

Comment:
This `enable_if_t` and the base specialization of `false_type` seems redundant with the `__is_equality_self_comparable<_Iterator1, _Iterator2>` inside the conjunction.  I think we can fully define it with the conjunction, and remove the enable if and base definition.

Reply:
Refactored.